### ISSUE TITLE
Cause Angular AoT build errors

### ISSUE THREAD

**Issue by kzanagodev** [moebius-mlm/ng-validators#3]
Hi,

This plugin cause an "AoT build failed" error when compiling a project in production mode.
Output of error : 

`typescript error:
         Error encountered resolving symbol values statically. Calling function 'ɵmakeDecorator', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, 
            resolving symbol Injectable in path_to_project/node_modules/@moebius/ng-validators/node_modules/@angular/core/core.d.ts,
            resolving symbol ɵi in path_to_project/node_modules/@moebius/ng-validators/node_modules/@angular/forms/forms.d.ts,
            resolving symbol ɵi in path_to_project/node_modules/@moebius/ng-validators/node_modules/@angular/forms/forms.d.ts

Error: The Angular AoT build failed.`

Your plugin seems to not been compatible with Angular AoT,

Thank you

**Comment by slavafomin** [comment]
Hello @kzanagodev!

Thank you for letting me know.

Do you know what exactly is causing this error?

We are using this module in production, without any compilation issues so far. What version of Angular do you use?

**Comment by kzanagodev** [comment]
Hi @slavafomin ,

I'm using this plugin in a ionic project.
Any issues when I run the project in dev mode but in production mode I got this AoT error.
The production mode is mandatory to reduce the app size and is launched with this command with Ionic :

 $ ionic cordova build android --prod

Environnement informations : 
    @ionic/cli-utils  : 1.19.0
    ionic (Ionic CLI) : 3.19.0
    cordova (Cordova CLI) : 8.0.0
    @ionic/app-scripts : 3.1.7
    Cordova Platforms  : android 6.3.0
    Ionic Framework    : ionic-angular 3.9.2
    Android SDK Tools : 26.0.2
    Node              : v6.10.3
    npm               : 5.6.0

Thank you

**Comment by slavafomin** [comment]
I'm not really familiar with ionic/cordova, will you be able to create a minimal project to reproduce the problem and push it to Github?

**Comment by darkbasic** [comment]
A simple `import {matchOtherValidator} from '@moebius/ng-validators';` is enough to trigger
```
ERROR in Error during template compile of 'ɵi'
  Function calls are not supported in decorators but 'ɵmakeDecorator' was called in 'Injectable'
    'Injectable' calls 'ɵmakeDecorator'.
```
in an AngularCLI project runned with `ng serve --aot`.

Also even in AOT mode I get the following when using `matchOtherValidator`:

```
WARNING in ./node_modules/@moebius/ng-validators/node_modules/@angular/core/@angular/core.es5.js
5659:15-36 Critical dependency: the request of a dependency is an expression
    at ImportLazyContextDependency.getWarnings (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/dependencies/ContextDependency.js:39:18)
    at Compilation.reportDependencyErrorsAndWarnings (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/Compilation.js:703:24)
    at Compilation.finish (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/Compilation.js:561:9)
    at applyPluginsParallel.err (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/Compiler.js:502:17)
    at /home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/tapable/lib/Tapable.js:289:11
    at /home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/html-webpack-plugin/index.js:60:9
    at tryCatcher (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:789:20)
    at tryOnImmediate (timers.js:751:5)
 @ ./node_modules/@moebius/ng-validators/node_modules/@angular/core/@angular/core.es5.js
 @ ./node_modules/@moebius/ng-validators/build/input-validators.module.js
 @ ./node_modules/@moebius/ng-validators/build/index.js
 @ ./src/app/login/containers/login.component.ts
 @ ./src/app/login/login.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

WARNING in ./node_modules/@moebius/ng-validators/node_modules/@angular/core/@angular/core.es5.js
5675:15-102 Critical dependency: the request of a dependency is an expression
    at ImportLazyContextDependency.getWarnings (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/dependencies/ContextDependency.js:39:18)
    at Compilation.reportDependencyErrorsAndWarnings (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/Compilation.js:703:24)
    at Compilation.finish (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/Compilation.js:561:9)
    at applyPluginsParallel.err (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/webpack/lib/Compiler.js:502:17)
    at /home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/tapable/lib/Tapable.js:289:11
    at /home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/html-webpack-plugin/index.js:60:9
    at tryCatcher (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/home/niko/WebstormProjects/new-tutorial/whatsapp-client-angularcli-material/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:789:20)
    at tryOnImmediate (timers.js:751:5)
 @ ./node_modules/@moebius/ng-validators/node_modules/@angular/core/@angular/core.es5.js
 @ ./node_modules/@moebius/ng-validators/build/input-validators.module.js
 @ ./node_modules/@moebius/ng-validators/build/index.js
 @ ./src/app/login/containers/login.component.ts
 @ ./src/app/login/login.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
```

```
signUpForm = this.fb.group({
    username: [null, [
      Validators.required,
    ]],
    newPassword: [null, [
      Validators.required,
    ]],
    confirmPassword: [null, [
      Validators.required,
      matchOtherValidator('password'),
    ]],
  });
```

**Comment by PitPanda1** [comment]
I have the same "problem" as @darkbasic (angular-cli 1.7, ng 5.2.5)

```
./node_modules/@moebius/ng-validators/node_modules/@angular/core/@angular/core.es5.js
5659:15-36 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/@moebius/ng-validators/node_modules/@angular/core/@angular/core.es5.js
 @ ./node_modules/@moebius/ng-validators/build/input-validators.module.js
 @ ./node_modules/@moebius/ng-validators/build/index.js
 @ ./src/app/backend/users/admin-users.component.ts
 @ ./src/app/backend/backend.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
    warnings @ client:153
    onmessage @ socket.js:41
    EventTarget.dispatchEvent @ sockjs.js:170
    (anonymous) @ sockjs.js:883
    SockJS._transportMessage @ sockjs.js:881
    EventEmitter.emit @ sockjs.js:86
    WebSocketTransport.ws.onmessage @ sockjs.js:2957
    wrapFn @ zone.js:1166
    ZoneDelegate.invokeTask @ zone.js:421
    Zone.runTask @ zone.js:188
    ZoneTask.invokeTask @ zone.js:496
    invokeTask @ zone.js:1517
    globalZoneAwareCallback @ zone.js:1543
```

I am using also the `matchOtherValidator`, but these warnings appear also when I comment them out. Maybe its just another problem with the `passwordValidator`, when commenting it out also, the warnings disappear. But they are the only validators in the whole project I use, maybe the warnings will come when using other validators too, didn't test it.

```
groupObj.prepassword = ['', [Validators.required, passwordValidator({
  minLength: 5,
  maxLength: 25,
  requireLetters: true,
  requireUpperCaseLetters: true,
  requireLowerCaseLetters: true,
  requireNumbers: true,
  requireSpecialCharacters: true
})]];
groupObj.password = ['', [Validators.required, matchOtherValidator('prepassword')]];
```

**Comment by Leobaillard** [comment]
I'm also experiencing this problem in an Ionic v3 app. Any news on this?

**Comment by slavafomin** [comment]
Hello guys!

Sorry it took me so long, but I've heavily updated the project and released version `2.0.0`. It is now built with `ng-packagr`, using best practices for Angular libraries. Also, it now supports the latest Angular version 8. Dependencies were also refactored.

I believe it should fix all the build issues. I'm going to close this issue for now, but feel free to let me know in case if you still have problems.
